packages/runtime/src/plugins/useUpstreamRetry.ts

```diff
@@ -0,0 +1,127 @@
+import { subgraphNameByExecutionRequest } from '@graphql-mesh/fusion-runtime';
+import { mapMaybePromise, MaybePromise } from '@graphql-tools/utils';
+import { DisposableSymbols } from '@whatwg-node/disposablestack';
+import { GatewayContext, GatewayPlugin } from '../types';
+
+export interface UpstreamRetryOptions {
+  /**
+   * The maximum number of retries to attempt.
+   */
+  maxRetries: number;
+  /**
+   * The delay between retries in milliseconds.
+   * If the upstream returns `Retry-After` header, the delay will be the value of the header.
+   * @default 1000
+   */
+  retryDelay?: number;
+  /**
+   * A function that determines whether a response should be retried.
+   * If the upstream returns `Retry-After` header, the response will be retried.
+   *
+   * @default (response) => response.status >= 500 || response.status === 429
+   */
+  shouldRetry?: (response: Response) => boolean;
```

Comment:
We should probably have a `payload: { response: Response }` here, so that we can easily add other metadata, such as the execution context, the subgraph name, etc...